File: templates/scenarios/react-vite/api/internal/testutil/modeltest/formal.go
package modeltest

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

type FormalArtifact struct {
	SchemaVersion int                        `json:"schemaVersion"`
	FlowID        string                     `json:"flowId"`
	Source        FormalArtifactSource       `json:"source"`
	Commands      map[string][]string        `json:"commands"`
	States        []string                   `json:"states"`
	Events        []string                   `json:"events"`
	Transitions   []FormalArtifactTransition `json:"transitions"`
	Traces        []FormalArtifactTrace      `json:"traces"`
	Checks        FormalArtifactChecks       `json:"checks"`
}

type FormalArtifactSource struct {
	ModelPath    string `json:"modelPath"`
	ModelSHA256  string `json:"modelSha256"`
	QuintVersion string `json:"quintVersion"`
}

type FormalArtifactChecks struct {
	Typechecked        bool `json:"typechecked"`
	Tested             bool `json:"tested"`
	Verified           bool `json:"verified"`
	GeneratedFromModel bool `json:"generatedFromModel"`
}

type FormalArtifactTransition struct {
	From      string `json:"from"`
	Event     string `json:"event"`
	To        string `json:"to"`
	WantError bool   `json:"wantError"`
}

type FormalArtifactTrace struct {
	Name    string                    `json:"name"`
	Initial string                    `json:"initial"`
	Steps   []FormalArtifactTraceStep `json:"steps"`
}

type FormalArtifactTraceStep struct {
	Event     string `json:"event"`
	Want      string `json:"want"`
	WantError bool   `json:"wantError"`
}

func LoadFormalArtifact(t TestingT, path string) FormalArtifact {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read formal artifact %s: %v", path, err)
	}
	var artifact FormalArtifact
	if err := json.Unmarshal(data, &artifact); err != nil {
		t.Fatalf("parse formal artifact %s: %v", path, err)
	}
	return artifact
}

func ValidateFormalArtifactFresh(artifact FormalArtifact, modelPath string) []error {
	var errs []error
	if artifact.SchemaVersion != 1 {
		errs = append(errs, fmt.Errorf("formal artifact schemaVersion=%d, want 1", artifact.SchemaVersion))
	}
	if artifact.FlowID == "" {
		errs = append(errs, fmt.Errorf("formal artifact flowId is required"))
	}
	if artifact.Source.ModelPath != modelPath {
		errs = append(errs, fmt.Errorf("formal artifact modelPath=%s, want %s", artifact.Source.ModelPath, modelPath))
	}
	if artifact.Source.QuintVersion == "" {
		errs = append(errs, fmt.Errorf("formal artifact quintVersion is required"))
	}
	if artifact.Source.ModelSHA256 == "" {
		errs = append(errs, fmt.Errorf("formal artifact modelSha256 is required"))
	} else {
		data, err := readModelFile(modelPath)
		if err != nil {
			errs = append(errs, fmt.Errorf("read formal model %s: %v", modelPath, err))
		} else {
			sum := sha256.Sum256(data)
			got := hex.EncodeToString(sum[:])
			if got != artifact.Source.ModelSHA256 {
				errs = append(errs, fmt.Errorf("formal artifact modelSha256=%s, want %s", artifact.Source.ModelSHA256, got))
			}
		}
	}
	if !artifact.Checks.Typechecked {
		errs = append(errs, fmt.Errorf("formal artifact was not typechecked"))
	}
	if !artifact.Checks.Tested {
		errs = append(errs, fmt.Errorf("formal artifact was not tested"))
	}
	if !artifact.Checks.Verified {
		errs = append(errs, fmt.Errorf("formal artifact was not verified"))
	}
	if !artifact.Checks.GeneratedFromModel {
		errs = append(errs, fmt.Errorf("formal artifact was not generated from model"))
	}
	if len(artifact.Transitions) == 0 {
		errs = append(errs, fmt.Errorf("formal artifact transitions must not be empty"))
	}
	if len(artifact.Traces) == 0 {
		errs = append(errs, fmt.Errorf("formal artifact traces must not be empty"))
	}
	return errs
}

func readModelFile(modelPath string) ([]byte, error) {
	var firstErr error
	candidates := []string{modelPath}
	if trimmed, ok := strings.CutPrefix(modelPath, "api/"); ok {
		candidates = append(candidates, trimmed)
	}
	candidates = append(candidates, filepath.Base(modelPath))

	wd, err := os.Getwd()
	if err != nil {
		wd = "."
	}
	for dir := wd; ; dir = filepath.Dir(dir) {
		for _, candidate := range candidates {
			data, err := os.ReadFile(filepath.Join(dir, candidate))
			if err == nil {
				return data, nil
			}
			if firstErr == nil {
				firstErr = err
			}
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			break
		}
	}
	return nil, firstErr
}

func AssertFormalArtifactFresh(t TestingT, artifact FormalArtifact, modelPath string) {
	t.Helper()
	if errs := ValidateFormalArtifactFresh(artifact, modelPath); len(errs) > 0 {
		t.Fatalf("formal artifact is stale or incomplete:\n%s", formatErrors(errs))
	}
}

func ValidateFormalTransitionsReplay[S comparable, E comparable](
	artifact FormalArtifact,
	statuses []S,
	events []E,
	transition Transition[S, E],
) []error {
	rows, errs := formalRows(artifact, statuses, events)
	if len(errs) > 0 {
		return errs
	}
	return ValidateTransitionMatrix(statuses, events, rows, transition)
}

func AssertFormalTransitionsReplay[S comparable, E comparable](
	t TestingT,
	artifact FormalArtifact,
	statuses []S,
	events []E,
	transition Transition[S, E],
) {
	t.Helper()
	if errs := ValidateFormalTransitionsReplay(artifact, statuses, events, transition); len(errs) > 0 {
		t.Fatalf("formal transition replay mismatch:\n%s", formatErrors(errs))
	}
}

func ValidateFormalTracesReplay[S comparable, E comparable](
	artifact FormalArtifact,
	statuses []S,
	events []E,
	transition Transition[S, E],
) []error {
	traces, errs := formalTraces(artifact, statuses, events)
	if len(errs) > 0 {
		return errs
	}
	return ValidateTraces(traces, transition)
}

func AssertFormalTracesReplay[S comparable, E comparable](
	t TestingT,
	artifact FormalArtifact,
	statuses []S,
	events []E,
	transition Transition[S, E],
) {
	t.Helper()
	if errs := ValidateFormalTracesReplay(artifact, statuses, events, transition); len(errs) > 0 {
		t.Fatalf("formal trace replay mismatch:\n%s", formatErrors(errs))
	}
}

func formalRows[S comparable, E comparable](
	artifact FormalArtifact,
	statuses []S,
	events []E,
) ([]MatrixRow[S, E], []error) {
	statusByName := valuesByString(statuses)
	eventByName := valuesByString(events)
	rows := make([]MatrixRow[S, E], 0, len(artifact.Transitions))
	var errs []error
	for i, transition := range artifact.Transitions {
		bad := false
		from, ok := statusByName[transition.From]
		if !ok {
			errs = append(errs, fmt.Errorf("formal transition %d unknown from state %s", i, transition.From))
			bad = true
		}
		to, ok := statusByName[transition.To]
		if !ok {
			errs = append(errs, fmt.Errorf("formal transition %d unknown to state %s", i, transition.To))
			bad = true
		}
		event, ok := eventByName[transition.Event]
		if !ok {
			errs = append(errs, fmt.Errorf("formal transition %d unknown event %s", i, transition.Event))
			bad = true
		}
		if bad {
			continue
		}
		rows = append(rows, MatrixRow[S, E]{
			Name:    fmt.Sprintf("formal transition %s/%s", transition.From, transition.Event),
			From:    from,
			Event:   event,
			To:      to,
			WantErr: transition.WantError,
		})
	}
	return rows, errs
}

func formalTraces[S comparable, E comparable](
	artifact FormalArtifact,
	statuses []S,
	events []E,
) ([]Trace[S, E], []error) {
	statusByName := valuesByString(statuses)
	eventByName := valuesByString(events)
	traces := make([]Trace[S, E], 0, len(artifact.Traces))
	var errs []error
	for traceIndex, formalTrace := range artifact.Traces {
		initial, ok := statusByName[formalTrace.Initial]
		if !ok {
			errs = append(errs, fmt.Errorf("formal trace %s unknown initial state %s", formalTrace.Name, formalTrace.Initial))
			continue
		}
		steps := make([]TraceStep[S, E], 0, len(formalTrace.Steps))
		for stepIndex, step := range formalTrace.Steps {
			bad := false
			event, ok := eventByName[step.Event]
			if !ok {
				errs = append(errs, fmt.Errorf("formal trace %s step %d unknown event %s", formalTrace.Name, stepIndex, step.Event))
				bad = true
			}
			want, ok := statusByName[step.Want]
			if !ok {
				errs = append(errs, fmt.Errorf("formal trace %s step %d unknown want state %s", formalTrace.Name, stepIndex, step.Want))
				bad = true
			}
			if bad {
				continue
			}
			steps = append(steps, TraceStep[S, E]{
				Name:    fmt.Sprintf("step %d", stepIndex),
				Event:   event,
				Want:    want,
				WantErr: step.WantError,
			})
		}
		traces = append(traces, Trace[S, E]{
			Name:    traceName(formalTrace.Name, traceIndex),
			Initial: initial,
			Steps:   steps,
		})
	}
	return traces, errs
}

func valuesByString[T comparable](values []T) map[string]T {
	byName := make(map[string]T, len(values))
	for _, value := range values {
		byName[fmt.Sprint(value)] = value
	}
	return byName
}

func traceName(name string, index int) string {
	if name != "" {
		return name
	}
	return fmt.Sprintf("formal trace %d", index)
}

File: scenarios/react-component-library/ui/src/test-utils/modeltest/formal.ts
import type { MatrixRow, WorkflowTransition } from "./matrix";
import { validateTransitionMatrix } from "./matrix";
import type { Trace } from "./traces";
import { validateTraces } from "./traces";

const formalArtifactSchemaVersion = 6;
const sha256Pattern = /^[a-f0-9]{64}$/;

export interface FormalArtifact {
  readonly schemaVersion: number;
  readonly flowId: string;
  readonly source: {
    readonly contractPath: string;
    readonly contractSha256: string;
    readonly generatorPath: string;
    readonly generatorSha256: string;
    readonly generatorVersion: number;
    readonly modelPath: string;
    readonly modelSha256: string;
    readonly quintVersion: string;
    readonly verificationBackend: string;
  };
  readonly commands: Record<string, readonly string[]>;
  readonly states: readonly string[];
  readonly events: readonly string[];
  readonly transitions: readonly FormalArtifactTransition[];
  readonly namedTraces: readonly FormalArtifactTrace[];
  readonly generatedTraces: readonly FormalArtifactTrace[];
  readonly invariants: readonly string[];
  readonly generatedChecks: readonly string[];
  readonly coverage: {
    readonly transitionMatrixComplete: boolean;
    readonly terminalTransitionsChecked: boolean;
    readonly namedTraces: FormalArtifactTraceCoverage;
    readonly generatedTraces: FormalArtifactTraceCoverage;
  };
  readonly checks: {
    readonly typechecked: boolean;
    readonly tested: boolean;
    readonly verified: boolean;
    readonly generatedFromContract: boolean;
    readonly generatedFromModel: boolean;
  };
}

export interface FormalArtifactTraceCoverage {
  readonly allStatesCovered: boolean;
  readonly allEventsCovered: boolean;
  readonly coveredStates: readonly string[];
  readonly coveredEvents: readonly string[];
  readonly coveredPairs?: readonly string[];
  readonly allPairsCovered?: boolean;
}

export interface FormalArtifactTransition {
  readonly from: string;
  readonly event: string;
  readonly to: string;
  readonly wantError: boolean;
}

export interface FormalArtifactTrace {
  readonly name: string;
  readonly initial: string;
  readonly steps: readonly FormalArtifactTraceStep[];
}

export interface FormalArtifactTraceStep {
  readonly event: string;
  readonly want: string;
  readonly wantError: boolean;
}

export interface FormalReplayAdapter<
  State extends PropertyKey,
  Event extends PropertyKey,
  RuntimeState,
  RuntimeEvent,
> {
  readonly states: readonly State[];
  readonly events: readonly Event[];
  readonly stateFor: Record<State, () => RuntimeState>;
  readonly eventFor: Record<Event, () => RuntimeEvent>;
  readonly statusOf: (state: RuntimeState) => State;
  readonly transition: (state: RuntimeState, event: RuntimeEvent) => RuntimeState;
}

export const transitionFromReplayAdapter =
  <State extends PropertyKey, Event extends PropertyKey, RuntimeState, RuntimeEvent>(
    adapter: FormalReplayAdapter<State, Event, RuntimeState, RuntimeEvent>,
  ): WorkflowTransition<State, Event> =>
  (state, event) =>
    adapter.statusOf(adapter.transition(adapter.stateFor[state](), adapter.eventFor[event]()));

export interface FormalArtifactFreshExpectation {
  readonly contractPath: string;
  readonly contractSha256?: string;
  readonly modelPath: string;
  readonly modelSha256?: string;
  readonly generatorPath?: string;
  readonly generatorSha256?: string;
  readonly invariants?: readonly string[];
  readonly generatedChecks?: readonly string[];
}

export const validateFormalArtifactFresh = (
  artifact: FormalArtifact,
  expected: FormalArtifactFreshExpectation,
): string[] => {
  const errors: string[] = [];
  if (artifact.schemaVersion !== formalArtifactSchemaVersion) {
    errors.push(
      `formal artifact schemaVersion=${artifact.schemaVersion}, want ${formalArtifactSchemaVersion}`,
    );
  }
  if (artifact.flowId.trim() === "") {
    errors.push("formal artifact flowId is required");
  }
  if (artifact.source.contractPath !== expected.contractPath) {
    errors.push(
      `formal artifact contractPath=${artifact.source.contractPath}, want ${expected.contractPath}`,
    );
  }
  requireSha256(errors, "contractSha256", artifact.source.contractSha256);
  if (expected.contractSha256 && artifact.source.contractSha256 !== expected.contractSha256) {
    errors.push(
      `formal artifact contractSha256=${artifact.source.contractSha256}, want ${expected.contractSha256}`,
    );
  }
  if (artifact.source.modelPath !== expected.modelPath) {
    errors.push(
      `formal artifact modelPath=${artifact.source.modelPath}, want ${expected.modelPath}`,
    );
  }
  if (expected.generatorPath && artifact.source.generatorPath !== expected.generatorPath) {
    errors.push(
      `formal artifact generatorPath=${artifact.source.generatorPath}, want ${expected.generatorPath}`,
    );
  }
  requireSha256(errors, "generatorSha256", artifact.source.generatorSha256);
  if (expected.generatorSha256 && artifact.source.generatorSha256 !== expected.generatorSha256) {
    errors.push(
      `formal artifact generatorSha256=${artifact.source.generatorSha256}, want ${expected.generatorSha256}`,
    );
  }
  if (!Number.isInteger(artifact.source.generatorVersion) || artifact.source.generatorVersion < 1) {
    errors.push("formal artifact generatorVersion is required");
  }
  if (artifact.source.verificationBackend.trim() === "") {
    errors.push("formal artifact verificationBackend is required");
  }
  if (artifact.source.quintVersion.trim() === "") {
    errors.push("formal artifact quintVersion is required");
  }
  requireSha256(errors, "modelSha256", artifact.source.modelSha256);
  if (expected.modelSha256 && artifact.source.modelSha256 !== expected.modelSha256) {
    errors.push(
      `formal artifact modelSha256=${artifact.source.modelSha256}, want ${expected.modelSha256}`,
    );
  }
  if (!artifact.checks.typechecked) {
    errors.push("formal artifact was not typechecked");
  }
  if (!artifact.checks.tested) {
    errors.push("formal artifact was not tested");
  }
  if (!artifact.checks.verified) {
    errors.push("formal artifact was not verified");
  }
  if (!artifact.checks.generatedFromContract) {
    errors.push("formal artifact was not generated from contract");
  }
  if (!artifact.checks.generatedFromModel) {
    errors.push("formal artifact was not generated from model");
  }
  for (const invariant of expected.invariants ?? []) {
    if (!artifact.invariants.includes(invariant)) {
      errors.push(`formal artifact missing invariant ${invariant}`);
    }
  }
  for (const check of expected.generatedChecks ?? []) {
    if (!artifact.generatedChecks.includes(check)) {
      errors.push(`formal artifact missing generated check ${check}`);
    }
  }
  if (!artifact.coverage.transitionMatrixComplete) {
    errors.push("formal artifact transition matrix is incomplete");
  }
  if (!artifact.coverage.terminalTransitionsChecked) {
    errors.push("formal artifact does not check terminal transitions");
  }
  if (!artifact.coverage.namedTraces.allStatesCovered) {
    errors.push("formal artifact named traces do not cover all states");
  }
  if (!artifact.coverage.namedTraces.allEventsCovered) {
    errors.push("formal artifact named traces do not cover all events");
  }
  if (artifact.coverage.generatedTraces.coveredStates.length === 0) {
    errors.push("formal artifact generated traces do not report covered states");
  }
  if (artifact.coverage.generatedTraces.coveredEvents.length === 0) {
    errors.push("formal artifact generated traces do not report covered events");
  }
  if (!artifact.coverage.generatedTraces.coveredPairs) {
    errors.push("formal artifact generated traces do not report covered pairs");
  }
  if (artifact.transitions.length === 0) {
    errors.push("formal artifact transitions must not be empty");
  }
  if (artifact.namedTraces.length === 0) {
    errors.push("formal artifact namedTraces must not be empty");
  }
  if (artifact.generatedTraces.length === 0) {
    errors.push("formal artifact generatedTraces must not be empty");
  }
  return errors;
};

const requireSha256 = (errors: string[], field: string, value: string): void => {
  if (!sha256Pattern.test(value)) {
    errors.push(`formal artifact ${field} is required`);
  }
};

export const assertFormalArtifactFresh = (
  artifact: FormalArtifact,
  expected: FormalArtifactFreshExpectation,
): void => {
  const errors = validateFormalArtifactFresh(artifact, expected);
  if (errors.length > 0) {
    throw new Error(`formal artifact is stale or incomplete:\n${formatErrors(errors)}`);
  }
};

export const validateFormalTransitionsReplay = <
  State extends PropertyKey,
  Event extends PropertyKey,
>(
  artifact: FormalArtifact,
  states: readonly State[],
  events: readonly Event[],
  transition: WorkflowTransition<State, Event>,
): string[] => {
  const { rows, errors } = formalRows(artifact, states, events);
  if (errors.length > 0) {
    return errors;
  }
  return validateTransitionMatrix(states, events, rows, transition);
};

export const assertFormalTransitionsReplay = <State extends PropertyKey, Event extends PropertyKey>(
  artifact: FormalArtifact,
  states: readonly State[],
  events: readonly Event[],
  transition: WorkflowTransition<State, Event>,
): void => {
  const errors = validateFormalTransitionsReplay(artifact, states, events, transition);
  if (errors.length > 0) {
    throw new Error(`formal transition replay mismatch:\n${formatErrors(errors)}`);
  }
};

export const validateFormalTracesReplay = <State extends PropertyKey, Event extends PropertyKey>(
  artifact: FormalArtifact,
  states: readonly State[],
  events: readonly Event[],
  transition: WorkflowTransition<State, Event>,
): string[] => {
  const { traces, errors } = formalTraces(
    [...artifact.namedTraces, ...artifact.generatedTraces],
    states,
    events,
  );
  if (errors.length > 0) {
    return errors;
  }
  return validateTraces(traces, transition);
};

export const assertFormalTracesReplay = <State extends PropertyKey, Event extends PropertyKey>(
  artifact: FormalArtifact,
  states: readonly State[],
  events: readonly Event[],
  transition: WorkflowTransition<State, Event>,
): void => {
  const errors = validateFormalTracesReplay(artifact, states, events, transition);
  if (errors.length > 0) {
    throw new Error(`formal trace replay mismatch:\n${formatErrors(errors)}`);
  }
};

const formalRows = <State extends PropertyKey, Event extends PropertyKey>(
  artifact: FormalArtifact,
  states: readonly State[],
  events: readonly Event[],
): { rows: MatrixRow<State, Event>[]; errors: string[] } => {
  const stateByName = valuesByString(states);
  const eventByName = valuesByString(events);
  const rows: MatrixRow<State, Event>[] = [];
  const errors: string[] = [];

  artifact.transitions.forEach((transition, index) => {
    const from = stateByName.get(transition.from);
    const to = stateByName.get(transition.to);
    const event = eventByName.get(transition.event);
    if (from === undefined) {
      errors.push(`formal transition ${index} unknown from state ${transition.from}`);
    }
    if (to === undefined) {
      errors.push(`formal transition ${index} unknown to state ${transition.to}`);
    }
    if (event === undefined) {
      errors.push(`formal transition ${index} unknown event ${transition.event}`);
    }
    if (from === undefined || to === undefined || event === undefined) {
      return;
    }
    rows.push({
      name: `formal transition ${transition.from}/${transition.event}`,
      from,
      event,
      to,
      wantError: transition.wantError,
    });
  });

  return { rows, errors };
};

const formalTraces = <State extends PropertyKey, Event extends PropertyKey>(
  artifactTraces: readonly FormalArtifactTrace[],
  states: readonly State[],
  events: readonly Event[],
): { traces: Trace<State, Event>[]; errors: string[] } => {
  const stateByName = valuesByString(states);
  const eventByName = valuesByString(events);
  const traces: Trace<State, Event>[] = [];
  const errors: string[] = [];

  artifactTraces.forEach((trace, traceIndex) => {
    const initial = stateByName.get(trace.initial);
    if (initial === undefined) {
      errors.push(`formal trace ${trace.name} unknown initial state ${trace.initial}`);
      return;
    }
    const steps = trace.steps.flatMap((step, stepIndex) => {
      const event = eventByName.get(step.event);
      const want = stateByName.get(step.want);
      if (event === undefined) {
        errors.push(`formal trace ${trace.name} step ${stepIndex} unknown event ${step.event}`);
      }
      if (want === undefined) {
        errors.push(`formal trace ${trace.name} step ${stepIndex} unknown want state ${step.want}`);
      }
      if (event === undefined || want === undefined) {
        return [];
      }
      return [{ event, want, wantError: step.wantError }];
    });
    traces.push({
      name: trace.name || `formal trace ${traceIndex}`,
      initial,
      steps,
    });
  });

  return { traces, errors };
};

const valuesByString = <Value extends PropertyKey>(values: readonly Value[]): Map<string, Value> =>
  new Map(values.map((value) => [String(value), value]));

const formatErrors = (errors: readonly string[]): string =>
  errors.map((error) => `  - ${error}`).join("\n");
